faculty/clients/api.py

```diff
@@ -0,0 +1,745 @@
+# Copyright 2018-2021 Faculty Science Limited
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Manage Faculty APIs.
+"""
+import uuid
+from datetime import datetime
+from enum import Enum
+from typing import List, Optional, Union
+
+from attr import define
+from marshmallow import fields, post_load
+from marshmallow_enum import EnumField
+
+from faculty._oneofschema import OneOfSchema
+from faculty.clients.base import BaseClient, BaseSchema
+from faculty.clients.job import InstanceSize, _InstanceSizeSchema
+from faculty.clients.server import Server, _ServerSchema
+
+
+class DeploymentStatus(Enum):
+    """An enumeration of possible API deployment statuses."""
+
+    NOTDEPLOYED = "not-deployed"
+    STARTING = "starting"
+    DEPLOYED = "deployed"
+    ERROR = "error"
+
+
+@define
```

Comment:
Rather than attrs new interface, I'd actually advocate for using dataclasses now that we're only supporting 3.6+ (dataclasses are 3.7+, but there is a 3.6 backport that [you can install on Python 3.6 only](https://github.com/acroz/pylivy/blob/master/setup.py#L28)). The syntax is basically the same but this has the obvious advantage of being in the standard library and better integrated with other Python tools.

I'd suggest we use dataclasses here and gradually migrate other clients to use dataclasses in future. What do you think?

Reply:
Yeah, I think this is a good call. Don't have strong reason to go with `attr`, not given how do we use things.
The plan sounds very sensible, will do the changes here, and once merged, will create an issue to track the whole migration as you outline.